### ISSUE TITLE
Adding SingleR cluster annotations back into Seurat Object

### ISSUE THREAD

**Issue by xAZx** [LTLA/SingleR#63]
Hi There!

I am currently attempting to add the SingleR cluster annotations back into my original Seurat Object (on which I've already run UMAP clustering and have 14 cell clusters), but unfortunately it does not seem to be working (the metadata column just outputs `<NA>`).

This is my current code:

 #convert Seurat object into SingleCellExperiment
`matrix <- as.SingleCellExperiment(p1_1)`

  #load immune cell database
`imm <- DatabaseImmuneCellExpressionData()`

 #Grab the common genes between the sets.
`common <- intersect(rownames(matrix), rownames(imm))
imm <- imm[common,]
matrix <- matrix[common,]`

 #Run SingleR
`singler <- SingleR(test = matrix, ref = imm, 
                   labels = imm$label.fine,
                   method = 'cluster', 
                   assay.type.ref = 'logcounts',
                   clusters = matrix$seurat_clusters)`

 #Add SingleR clusters back into Seurat object
`p1_1[['singler_clusters']] <- singler$labels[match(p1_1[[]][['seurat_clusters']],
                                                   singler$first.labels)]`

I believe the issue is with the last part where I am trying to add the SingleR assigned clusters back into my original Seurat object. The readme file has this piece of code:

`seurat.obj[["SingleR.cluster.labels"]] <- 
        singler.results$labels[match(seurat.obj[[]][["my.input.clusters"]], singler.results$clusts)]]`

But when I look at my 'singler' object, I have no 'clusts' variable (which is why I just used 'first.labels'). And once again, after I run the above code, my newly defined 'singler_clusters' metadata column is blank!

Any help would be greatly appreciated! Many thanks.

**Comment by LTLA** [comment]
Y'know, I can smell the panic in your code as you tried different things that didn't work. 😄 

It's good to take a deep breath to think about it. For starters, your cluster names will never be the same as the labels (otherwise, what's the point of annotating?) so putting in `first.labels` doesn't make any sense. The presence of `NA`s is subsequently a logical outcome of doing that, because when `match()` can't find a match, it will correctly emit `NA` values. 

If you have a look at `singler.results`, you will instead see that the names of your input clusters are embedded in the row names. So the functional code would look something like:

```r
seurat.obj[["SingleR.cluster.labels"]] <- singler.results$labels[match(seurat.obj[[]][["my.input.clusters"]], rownames(singler.results))]]
```

@j-andrews7 We should probably fix that example in the `README`, it must have been from the old version where there was a `clusts` field (presumably, I don't remember).

**P.S.** What on earth is `p1_1[[]]`? Is that the recommended way to access the metadata slots of a Seurat object? If so... woah. That's a 🇳🇿 for sure.

**Comment by j-andrews7** [comment]
That particular typo is fixed in #59 - I can't remember if it used to be that way or was a legit typo.

The `[[]]` malarkey is to get the cluster labels from the metadata as a vector rather than a dataframe for `match`. The empty `[[]]` returns the metadata as a dataframe, the additional `[["my.input.clusters]]` returns the appropriate column as a vector. I agree it's pretty gross - there may be a prettier way, but I couldn't think of one off the top of my head.

**Comment by xAZx** [comment]
Many thanks for the quick response! It worked perfectly, and I appreciate the explanation behind the code!
